### PR TITLE
Phase 3: add zeroed MASP data for all tokens

### PR DESCRIPTION
**Pull request by brentstone**

### CODE REVIEW

File: phase3/src/lib.rs [hunk 1]
```diff
@@ -28,14 +29,21 @@ const IBC_TOKENS: [(
...
+    // Read the current gas cost map
     let gas_cost_key = get_gas_cost_key();
     let mut minimum_gas_price: BTreeMap<Address, token::Amount> =
         ctx.read(&gas_cost_key)?.unwrap_or_default();
 
+    // Read the current MASP token map
+    let token_map_key = token::storage_key::masp_token_map_key();
+    let mut token_map = ctx
+        .read::<masp::TokenMap>(&token_map_key)?
+        .unwrap_or_default();
```
Comment (sug0) [inline phase3/src/lib.rs:41]: there's a `read_token_map` function in `shielded_token` you can use instead of this

File: phase3/src/lib.rs [hunk 2]
```diff
@@ -45,11 +53,43 @@ fn apply_tx(ctx: &mut Ctx, _tx_data: BatchedTx) -> TxResult {
...
+        )?;
+        ctx.write(&shielded_token_kp_gain_key, Dec::zero())?;
+        ctx.write(&shielded_token_kd_gain_key, Dec::zero())?;
     }
 
+    // Write the gas cost map back to storage
     ctx.write(&gas_cost_key, minimum_gas_price)?;
 
+    // Write the token map back to storage
+    ctx.write(&token_map_key, token_map)?;
```
Comment (sug0) [inline phase3/src/lib.rs:92]: also a `write_token_map` function

File: phase3/src/lib.rs [hunk 3]
```diff
@@ -45,11 +53,43 @@ fn apply_tx(ctx: &mut Ctx, _tx_data: BatchedTx) -> TxResult {
...
+            token::Amount::zero(),
+        )?;
+        ctx.write(&shielded_token_last_inflation_key, token::Amount::zero())?;
+        ctx.write(&shielded_token_max_rewards_key, Dec::zero())?;
+        ctx.write(
+            &shielded_token_target_locked_amount_key,
+            token::Amount::zero(),
+        )?;
+        ctx.write(&shielded_token_kp_gain_key, Dec::zero())?;
+        ctx.write(&shielded_token_kd_gain_key, Dec::zero())?;
```
Comment (sug0) [inline phase3/src/lib.rs:85]: `namada_shielded_token::write_params` takes care of all of this

**Review by sug0 — CHANGES_REQUESTED** [review]

**Comment by brentstone** [comment]
@tzemanovic we can't implement the suggested changes here from @sug0 because the shielded token crate is not exposed in the `tx_prelude`, should we make those changes to `namada` or just approve this PR as is?

**Comment by tzemanovic** [comment]
> @tzemanovic we can't implement the suggested changes here from @sug0 because the shielded token crate is not exposed in the `tx_prelude`, should we make those changes to `namada` or just approve this PR as is?

I'd say both - we should expose it but merge this in the meantime

**Pull request merged**